### PR TITLE
Fix Tarjan implementation

### PR DESCRIPTION
**Pull request by kortschak**
The previous code was giving incorrect results for a 2SAT solver I have
built on it (cannot release). This fixes that.

Tests to come.

### CODE REVIEW

File: search/graph_search.go [hunk 1]
```diff
@@ -382,59 +381,85 @@ func CopyDirectedGraph(dst graph.MutableDirectedGraph, src graph.DirectedGraph)
...
+		}
+	}
+	return t.sccs
+}
 
-	var strongconnect func(graph.Node) []graph.Node
+// tarjan implements Tarjan's strongly connected component finding
+// algorithm. The implementation is from the pseudocode at
+//
+// http://en.wikipedia.org/wiki/Tarjan%27s_strongly_connected_components_algorithm
```
Comment (soniakeys) [inline search/graph_search.go]: It would be good to add a date accessed as part of the reference here.  Wikipedia changes.

**Comment by soniakeys** [comment]
Ugh.  The current WP pseudocode doesn't even look correct.  The commit comment makes the assumption that it's expensive to check if a node is on the stack and introduces the isRemoved flag.  But without initialization, it's unclear that !isRemoved is a valid way to check if a node is on the stack.  Instead of isRemoved, I would add a an "onStack" flag.

**Comment by kortschak** [comment]
I have made that change. PTAL

Would you edit the WP article, or would you like me to add a comment that it is with changes? I'll update the comment when I know which.

**Comment by kortschak** [comment]
*ping*

**Comment by soniakeys** [comment]
Sorry I was distracted with other things for a bit.  LGTM.  For the comment, I would reference the previous version at WP with "http://en.wikipedia.org/w/index.php?title=Tarjan%27s_strongly_connected_components_algorithm&oldid=642744644."  Pseudocode in that version simply reads `if (w is in S)` without attempting to direct implementation too much.  I think it's fine to indicate our code is based on that.

I had a WP account once.  You're right I should try to recover it and suggest changes there.

File: search/graph_search.go (hunk 1 above)
Comment (kortschak) [inline search/graph_search.go]: Fixed, thank you.

**Pull request merged**